File: dlitscustom/report/dlits_supplier_ledger_report/dlits_supplier_ledger_report.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

frappe.query_reports["DLITS Supplier Ledger Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "supplier",
            "label": __("Supplier"),
            "fieldtype": "Link",
            "options": "Supplier"
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "outstanding_amount" && data && flt(data.outstanding_amount) > 0) {
            value = "<span style='color:red'>" + value + "</span>";
        }
        
        if (column.fieldname == "aging_over_90" && data && flt(data.aging_over_90) > 0) {
            value = "<span style='color:red; font-weight:bold'>" + value + "</span>";
        }
        
        return value;
    },
    
    "onload": function(report) {
        // Add PDF Export button
        report.page.add_inner_button(__("Export PDF"), function() {
            let filters = report.get_values();
            let data = report.data || [];
            generate_pdf_report(data, filters, "DLITS Supplier Ledger Report");
        });
        
        // Add Print button
        report.page.add_inner_button(__("Print"), function() {
            let filters = report.get_values();
            let data = report.data || [];
            generate_pdf_report(data, filters, "DLITS Supplier Ledger Report");
        });
    }
};

function get_filter_string(filters) {
    let filter_array = [];
    
    if (filters.from_date) {
        filter_array.push("From: " + frappe.datetime.str_to_user(filters.from_date));
    }
    if (filters.to_date) {
        filter_array.push("To: " + frappe.datetime.str_to_user(filters.to_date));
    }
    if (filters.supplier) {
        filter_array.push("Supplier: " + filters.supplier);
    }
    
    return filter_array.join(" | ");
}

function generate_pdf_report(data, filters, report_name) {
    let filter_string = get_filter_string(filters);
    
    let html_content = `
        <div style="text-align: center; margin-bottom: 30px;">
            <h2 style="margin: 0; color: #333;">${report_name}</h2>
            <p style="margin: 5px 0; color: #666;">${filter_string}</p>
            <p style="margin: 5px 0; color: #666;">Generated on: ${frappe.datetime.str_to_user(frappe.datetime.get_today())}</p>
        </div>
        
        <table style="width: 100%; border-collapse: collapse; font-size: 12px;">
            <thead>
                <tr style="background-color: #f8f9fa;">
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: left;">Supplier</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Total Billed</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Total Paid</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Outstanding</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">0-30 Days</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">31-60 Days</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">61-90 Days</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Over 90 Days</th>
                </tr>
            </thead>
            <tbody>`;
    
    let total_billed = 0, total_paid = 0, total_outstanding = 0;
    let total_0_30 = 0, total_31_60 = 0, total_61_90 = 0, total_over_90 = 0;
    
    data.forEach(function(row) {
        total_billed += flt(row.total_billed);
        total_paid += flt(row.total_paid);
        total_outstanding += flt(row.outstanding_amount);
        total_0_30 += flt(row.aging_0_30);
        total_31_60 += flt(row.aging_31_60);
        total_61_90 += flt(row.aging_61_90);
        total_over_90 += flt(row.aging_over_90);
        
        html_content += `
            <tr>
                <td style="border: 1px solid #ddd; padding: 6px;">${row.supplier_name || row.supplier}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.total_billed)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.total_paid)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; ${flt(row.outstanding_amount) > 0 ? 'color: red; font-weight: bold;' : ''}">${format_currency(row.outstanding_amount)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.aging_0_30)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.aging_31_60)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.aging_61_90)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; ${flt(row.aging_over_90) > 0 ? 'color: red; font-weight: bold;' : ''}">${format_currency(row.aging_over_90)}</td>
            </tr>`;
    });
    
    // Add totals row
    html_content += `
            <tr style="background-color: #f8f9fa; font-weight: bold;">
                <td style="border: 1px solid #ddd; padding: 6px;">TOTAL</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_billed)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_paid)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; color: red;">${format_currency(total_outstanding)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_0_30)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_31_60)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_61_90)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; color: red;">${format_currency(total_over_90)}</td>
            </tr>
        </tbody>
    </table>
    
    <div style="margin-top: 30px; font-size: 11px; color: #666;">
        <p><strong>Summary:</strong></p>
        <ul>
            <li>Total Suppliers: ${data.length}</li>
            <li>Total Outstanding: ${format_currency(total_outstanding)}</li>
            <li>Overdue (>90 days): ${format_currency(total_over_90)}</li>
        </ul>
    </div>`;
    
    // Create and download PDF
    let pdf_window = window.open('', '_blank');
    pdf_window.document.write(`
        <html>
            <head>
                <title>${report_name}</title>
                <style>
                    body { font-family: Arial, sans-serif; margin: 20px; }
                    @media print { body { margin: 0; } }
                </style>
            </head>
            <body>
                ${html_content}
                <script>
                    window.onload = function() {
                        window.print();
                    }
                </script>
            </body>
        </html>
    `);
    pdf_window.document.close();
}

function format_currency(value) {
    return frappe.format(flt(value), {fieldtype: 'Currency'});
}

File: dlitscustom/report/dlits_supplier_ledger_report/dlits_supplier_ledger_report.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate, add_days
from datetime import datetime, timedelta

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data, filters)
    
    return columns, data, None, chart

def get_columns():
    return [
        {
            "fieldname": "supplier",
            "label": _("Supplier"),
            "fieldtype": "Link",
            "options": "Supplier",
            "width": 150
        },
        {
            "fieldname": "supplier_name",
            "label": _("Supplier Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "total_billed",
            "label": _("Total Billed"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "total_paid",
            "label": _("Total Paid"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "outstanding_amount",
            "label": _("Outstanding"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "aging_0_30",
            "label": _("0-30 Days"),
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "fieldname": "aging_31_60",
            "label": _("31-60 Days"),
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "fieldname": "aging_61_90",
            "label": _("61-90 Days"),
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "fieldname": "aging_over_90",
            "label": _("Over 90 Days"),
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "fieldname": "recent_bills",
            "label": _("Recent Bills"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "recent_payments",
            "label": _("Recent Payments"),
            "fieldtype": "Data",
            "width": 200
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    
    # Get supplier data with purchase invoices and payments
    query = """
        SELECT 
            s.name as supplier,
            s.supplier_name,
            COALESCE(SUM(pi.grand_total), 0) as total_billed,
            COALESCE(SUM(pi.paid_amount), 0) as total_paid,
            COALESCE(SUM(pi.outstanding_amount), 0) as outstanding_amount
        FROM `tabSupplier` s
        LEFT JOIN `tabPurchase Invoice` pi ON s.name = pi.supplier 
            AND pi.docstatus = 1 {conditions}
        GROUP BY s.name, s.supplier_name
        HAVING total_billed > 0 OR outstanding_amount > 0
        ORDER BY outstanding_amount DESC, total_billed DESC
    """.format(conditions=conditions)
    
    data = frappe.db.sql(query, filters, as_dict=True)
    
    # Add aging analysis and recent transactions
    for row in data:
        add_aging_analysis(row, filters)
        add_recent_transactions(row, filters)
    
    return data

def get_conditions(filters):
    conditions = ""
    
    if filters.get("from_date"):
        conditions += " AND pi.posting_date >= %(from_date)s"
    
    if filters.get("to_date"):
        conditions += " AND pi.posting_date <= %(to_date)s"
    
    if filters.get("supplier"):
        conditions += " AND s.name = %(supplier)s"
    
    return conditions

def add_aging_analysis(row, filters):
    """Add aging analysis for outstanding amounts"""
    supplier = row.get("supplier")
    to_date = filters.get("to_date") or frappe.utils.today()
    
    aging_query = """
        SELECT 
            SUM(CASE 
                WHEN DATEDIFF(%(to_date)s, pi.posting_date) <= 30 
                THEN pi.outstanding_amount ELSE 0 
            END) as aging_0_30,
            SUM(CASE 
                WHEN DATEDIFF(%(to_date)s, pi.posting_date) BETWEEN 31 AND 60 
                THEN pi.outstanding_amount ELSE 0 
            END) as aging_31_60,
            SUM(CASE 
                WHEN DATEDIFF(%(to_date)s, pi.posting_date) BETWEEN 61 AND 90 
                THEN pi.outstanding_amount ELSE 0 
            END) as aging_61_90,
            SUM(CASE 
                WHEN DATEDIFF(%(to_date)s, pi.posting_date) > 90 
                THEN pi.outstanding_amount ELSE 0 
            END) as aging_over_90
        FROM `tabPurchase Invoice` pi
        WHERE pi.supplier = %(supplier)s 
            AND pi.docstatus = 1 
            AND pi.outstanding_amount > 0
    """
    
    aging_data = frappe.db.sql(aging_query, {
        "supplier": supplier,
        "to_date": to_date
    }, as_dict=True)
    
    if aging_data:
        aging = aging_data[0]
        row.update({
            "aging_0_30": flt(aging.get("aging_0_30", 0)),
            "aging_31_60": flt(aging.get("aging_31_60", 0)),
            "aging_61_90": flt(aging.get("aging_61_90", 0)),
            "aging_over_90": flt(aging.get("aging_over_90", 0))
        })

def add_recent_transactions(row, filters):
    """Add recent bills and payments information"""
    supplier = row.get("supplier")
    
    # Recent bills
    recent_bills = frappe.db.sql("""
        SELECT name, posting_date, grand_total
        FROM `tabPurchase Invoice`
        WHERE supplier = %(supplier)s AND docstatus = 1
        ORDER BY posting_date DESC
        LIMIT 3
    """, {"supplier": supplier}, as_dict=True)
    
    bill_list = []
    for bill in recent_bills:
        bill_list.append(f"{bill.name} ({formatdate(bill.posting_date)}: {frappe.format_value(bill.grand_total, 'Currency')})")
    
    row["recent_bills"] = "; ".join(bill_list) if bill_list else "No recent bills"
    
    # Recent payments
    recent_payments = frappe.db.sql("""
        SELECT pe.name, pe.posting_date, pe.paid_amount
        FROM `tabPayment Entry` pe
        WHERE pe.party = %(supplier)s 
            AND pe.party_type = 'Supplier'
            AND pe.docstatus = 1
        ORDER BY pe.posting_date DESC
        LIMIT 3
    """, {"supplier": supplier}, as_dict=True)
    
    payment_list = []
    for pay in recent_payments:
        payment_list.append(f"{pay.name} ({formatdate(pay.posting_date)}: {frappe.format_value(pay.paid_amount, 'Currency')})")
    
    row["recent_payments"] = "; ".join(payment_list) if payment_list else "No recent payments"

def get_chart_data(data, filters):
    """Generate chart data for supplier ledger analysis"""
    if not data:
        return None
    
    # Aging analysis chart
    aging_labels = ["0-30 Days", "31-60 Days", "61-90 Days", "Over 90 Days"]
    aging_values = [0, 0, 0, 0]
    
    for row in data:
        aging_values[0] += flt(row.get("aging_0_30", 0))
        aging_values[1] += flt(row.get("aging_31_60", 0))
        aging_values[2] += flt(row.get("aging_61_90", 0))
        aging_values[3] += flt(row.get("aging_over_90", 0))
    
    return {
        "data": {
            "labels": aging_labels,
            "datasets": [{
                "name": "Outstanding Amount",
                "values": aging_values
            }]
        },
        "type": "donut",
        "height": 300,
        "colors": ["#28a745", "#ffc107", "#fd7e14", "#dc3545"]
    }